core/services/llo/delegate.go

```diff
@@ -0,0 +1,119 @@
+package llo
+
+import (
+	"context"
+	"fmt"
+
+	ocrcommontypes "github.com/smartcontractkit/libocr/commontypes"
+	ocr2types "github.com/smartcontractkit/libocr/offchainreporting2/types"
+	ocr2plus "github.com/smartcontractkit/libocr/offchainreporting2plus"
+	ocr3types "github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-data-streams/llo"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+	"github.com/smartcontractkit/chainlink/v2/core/services/llo/evm"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+	"github.com/smartcontractkit/chainlink/v2/core/services/streams"
+)
+
+var _ job.ServiceCtx = &delegate{}
+
+type Closer interface {
+	Close() error
+}
+
+type delegate struct {
+	services.StateMachine
+
+	cfg    DelegateConfig
+	codecs map[commontypes.LLOReportFormat]llo.ReportCodec
+
+	oracle Closer
+}
+
+type DelegateConfig struct {
+	Logger   logger.Logger
+	Queryer  pg.Queryer
+	Runner   streams.Runner
+	Registry Registry
+
+	// LLO
+	ChannelDefinitionCache commontypes.ChannelDefinitionCache
+
+	// OCR3
+	BinaryNetworkEndpointFactory ocr2types.BinaryNetworkEndpointFactory
+	V2Bootstrappers              []ocrcommontypes.BootstrapperLocator
+	ContractConfigTracker        ocr2types.ContractConfigTracker
+	ContractTransmitter          ocr3types.ContractTransmitter[commontypes.LLOReportInfo]
+	Database                     ocr3types.Database
+	OCRLogger                    ocrcommontypes.Logger
+	MonitoringEndpoint           ocrcommontypes.MonitoringEndpoint
+	OffchainConfigDigester       ocr2types.OffchainConfigDigester
+	OffchainKeyring              ocr2types.OffchainKeyring
+	OnchainKeyring               ocr3types.OnchainKeyring[commontypes.LLOReportInfo]
+	LocalConfig                  ocr2types.LocalConfig
+}
+
+func NewDelegate(cfg DelegateConfig) job.ServiceCtx {
+	if cfg.Queryer == nil {
+		panic("Queryer must not be nil")
```

Comment:
why panic instead of return an error?

Reply:
Figured it was most likely a programming error to fail to set this. One of the problems of using a config struct instead of positional args is that ppl can forget to include things. This at least enforces it gets set at runtime, which is not as good as enforcing at compile time, but its still something.

If we don't panic here, we panic further down the stack with a much more opaque error.

Reply:
I don't follow why it would have to be a panic instead of an error, here or down the stack.

What about returning an error from Start()? Or adding an error return here?

Reply:
It returns error now